src/Corax/IndexWriter.cs

```diff
@@ -1829,177 +1838,426 @@ private void InsertSpatialField(Tree entriesToSpatialTree, IndexedField indexedF
+                    _sortedTerms[idx] = k;
+                    _termIndexes[idx] = v;
+                    idx++;
+                }
+
+                terms = new Span<Slice>(_sortedTerms, 0, termsCount);
+                indexes = new Span<int>(_termIndexes, 0, termsCount);
+
+                terms.Sort(indexes, SliceComparer.Instance);
+            }
+
+            public TextualFieldBuffers(IndexWriter parent)
             {
-                if (sortedTermsBuffer.Length > 0)
-                    ArrayPool<Slice>.Shared.Return(sortedTermsBuffer);
-                sortedTermsBuffer = ArrayPool<Slice>.Shared.Rent(termsCount);
+                _parent = parent;
+                Keys = ArrayPool<CompactTree.CompactKeyLookup>.Shared.Rent(BatchSize);
+                PageOffsets = ArrayPool<int>.Shared.Rent(BatchSize);
+                PostListIds = ArrayPool<long>.Shared.Rent(BatchSize);
+                _entriesOffsets = ArrayPool<int>.Shared.Rent(BatchSize);
             }
 
-            currentFieldTerms.Keys.CopyTo(sortedTermsBuffer, 0);
+            public void Dispose()
+            {
+                if (PostListIds != null) ArrayPool<long>.Shared.Return(PostListIds);
+                if (PageOffsets != null) ArrayPool<int>.Shared.Return(PageOffsets);
+                if (_entriesOffsets != null) ArrayPool<int>.Shared.Return(_entriesOffsets);
 
-            // Sorting the terms buffer.
-            Sorter<Slice, SliceStructComparer> sorter = default;
-            sorter.Sort(sortedTermsBuffer, 0, termsCount);
+                if (_sortedTerms != null) ArrayPool<Slice>.Shared.Return(_sortedTerms);
+                if (_termIndexes != null) ArrayPool<int>.Shared.Return(_termIndexes);
+                
+                if (Keys != null)
+                {
+                    var llt = _parent._transaction.LowLevelTransaction;
+                    for (int i = 0; i < Keys.Length; i++)
+                    {
+                        ref var k = ref Keys[i].Key;
+                        if (k != null)
+                        {
+                           llt.ReleaseCompactKey(ref k); 
+                        }
+                    }
+                    ArrayPool<CompactTree.CompactKeyLookup>.Shared.Return(Keys);
+                }
 
-            using var dumper = new IndexTermDumper(_fieldsTree, indexedField.Name);
+                PostListIds = null;
+                PageOffsets = null;
+                _entriesOffsets = null;
+                Keys = null;
+            }
+        }
 
-            fieldTree.InitializeStateForTryGetNextValue();
-            long totalLengthOfTerm = 0;
+        private TextualFieldBuffers _textualFieldBuffers;
 
-            using var compactKeyCacheScope = new CompactKeyCacheScope(_transaction.LowLevelTransaction);
 
-            var entriesForTerm = new NativeList<TermInEntryModification>();
-            entriesForTerm.Initialize(_entriesAllocator);
+        private ref struct TextualFieldInserter
```

Comment:
Can we move it to separate file?

Reply:
[Same comment as above](https://github.com/ravendb/ravendb/pull/17176#discussion_r1315442938)